miden-tx/src/host/mod.rs

```diff
@@ -274,4 +282,36 @@ impl<A: AdviceProvider> Host for TransactionHost<A> {
 
         Ok(HostResponse::None)
     }
+
+    fn on_trace<S: ProcessState>(
+        &mut self,
+        process: &S,
+        trace_id: u32,
+    ) -> Result<HostResponse, ExecutionError> {
+        let event = TransactionTrace::try_from(trace_id)
+            .map_err(|err| ExecutionError::EventError(err.to_string()))?;
+
+        use TransactionTrace::*;
+        match event {
+            PrologueStart => self.tx_progress.start_prologue(process.clk()),
+            PrologueEnd => self.tx_progress.end_prologue(process.clk()),
+            NotesProcessingStart => self.tx_progress.start_notes_processing(process.clk()),
+            NotesProcessingEnd => self.tx_progress.end_notes_processing(process.clk()),
+            NoteExecutionStart => {
+                let note_id = get_current_note_id(process)?;
+                self.tx_progress.start_note_execution(process.clk(), note_id);
+            },
```

Comment:
Related to one of the previous comments: I'd probably make `get_current_note_id()` be a private method on `TransactionHost`.